Error while processing bom: Duplicate key

### Current Behavior

Hi, 
Our process makes daily scans of all our projects, create the SBOM and load it to Dependency Track, Unfortunately, for some projects we are getting errors like this:

`ERROR [BomUploadProcessingTask] Error while processing bom
java.lang.IllegalStateException: Duplicate key Identity[group=aquasecurity, name=trivy:PkgID, value=System.ValueTuple@4.5.0] (attempted merging values ComponentProperty{id=1450648, component=pkg:nuget/System.ValueTuple@4.5.0, groupName=aquasecurity, propertyName=trivy:PkgID, propertyValue=System.ValueTuple@4.5.0, propertyType=STRING, uuid=5e7640f8-2e76-4a09-8817-86cba50e0e05} and ComponentProperty{id=1450649, component=pkg:nuget/System.ValueTuple@4.5.0, groupName=aquasecurity, propertyName=trivy:PkgID, propertyValue=System.ValueTuple@4.5.0, propertyType=STRING, uuid=701f26af-fc1f-407e-966c-270c7e7325e8})
	at java.base/java.util.stream.Collectors.duplicateKeyException(Unknown Source)
	at java.base/java.util.stream.Collectors.lambda$uniqKeysMapAccumulator$1(Unknown Source)
	at java.base/java.util.stream.ReduceOps$3ReducingSink.accept(Unknown Source)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.copyInto(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source)
	at java.base/java.util.stream.AbstractPipeline.evaluate(Unknown Source)
	at java.base/java.util.stream.ReferencePipeline.collect(Unknown Source)
	at org.dependencytrack.persistence.ComponentQueryManager.synchronizeComponentProperties(ComponentQueryManager.java:925)
	at org.dependencytrack.persistence.QueryManager.synchronizeComponentProperties(QueryManager.java:608)
	at org.dependencytrack.parser.cyclonedx.util.ModelConverter.convert(ModelConverter.java:613)
	at org.dependencytrack.parser.cyclonedx.util.ModelConverter.convertComponents(ModelConverter.java:464)
	at org.dependencytrack.tasks.BomUploadProcessingTask.inform(BomUploadProcessingTask.java:157)
	at org.dependencytrack.tasks.BomUploadProcessingTaskV2.inform(BomUploadProcessingTaskV2.java:151)
	at alpine.event.framework.BaseEventService.lambda$publish$0(BaseEventService.java:110)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)`

Already there was one System.ValueTuple loaded in version 4.5.0:
![2024-08-01 16_58_45-Window](https://github.com/user-attachments/assets/8100c524-50e7-441d-ac86-f497882094a5)

In SBOM there was 4 separate instances of this component:
`    {
      "type": "library",
      "bom-ref": "684a5aa1-535f-4fd2-ba87-3ca30b505e7c",
      "name": "System.ValueTuple",
      "version": "4.5.0",
      "purl": "pkg:nuget/System.ValueTuple@4.5.0",
      "properties": [
        {
          "name": "aquasecurity:trivy:PkgType",
          "value": "nuget"
        }
      ]
    },
	    },
    {
      "type": "library",
      "bom-ref": "a895ab7c-cee3-414b-8811-0c7a9f01b968",
      "name": "System.ValueTuple",
      "version": "4.5.0",
      "purl": "pkg:nuget/System.ValueTuple@4.5.0",
      "properties": [
        {
          "name": "aquasecurity:trivy:PkgType",
          "value": "nuget"
        }
      ]
    },
	    {
      "type": "library",
      "bom-ref": "b4ace403-0b23-407c-9f9f-9a103de5d6f8",
      "name": "System.ValueTuple",
      "version": "4.5.0",
      "purl": "pkg:nuget/System.ValueTuple@4.5.0",
      "properties": [
        {
          "name": "aquasecurity:trivy:PkgType",
          "value": "nuget"
        }
      ]
    },
	    {
      "type": "library",
      "bom-ref": "0ddf76c9-bc51-4287-81b8-4a7690cfde09",
      "name": "System.ValueTuple",
      "version": "4.5.0",
      "purl": "pkg:nuget/System.ValueTuple@4.5.0",
      "properties": [
        {
          "name": "aquasecurity:trivy:PkgType",
          "value": "nuget"
        }
      ]
    },`

I had the same issue as per other issue, that the API/UI responds with 200 okay response to uploading this SBOM.

Loading this SBOM to empty project looks fine:
![2024-08-01 17_04_02-Dependency-Track - biomedical-knowledge-base_bkb-r-client ▸ 1 - Work - Microsoft](https://github.com/user-attachments/assets/168df4ed-5c3a-4de0-88ba-24d31c455314)

Any ideas how to go over that?



### Steps to Reproduce

1. One SBOM was loaded to project with one System.ValueTuple in 4.5.0 version defined.
2. New SBOM was loaded to project, 200 message was shown, but the SBOM is not loading with the error described above.

### Expected Behavior

The SBOM should be correctly replaced/updated.

### Dependency-Track Version

4.11.4

### Dependency-Track Distribution

Container Image

### Database Server

PostgreSQL

### Database Server Version

13.14

### Browser

Microsoft Edge

### Checklist

- [X] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/master/CONTRIBUTING.md#filing-issues)
- [X] I have checked the [existing issues](https://github.com/DependencyTrack/dependency-track/issues) for whether this defect was already reported

I have a similar issue, yet the stack trace of the exception specifically refers to the `LicenseQueryManager`:

```
2024-08-02 07:43:45,014 INFO [BomUploadProcessingTaskV2] Consuming uploaded BOM [bomSerialNumber=08e08436-c806-4460-85b9-301221203475, bomFormat=CycloneDX, bomUploadToken=f451988e-4d54-4f3e-9555-d3853f07ff19, projectName=__redacted__, bomSpecVersion=1.6, projectUuid=4e0d24b4-b3f1-45a2-9489-c604fc0b4aa6, projectVersion=__redacted__, bomVersion=1]
2024-08-02 07:43:45,029 INFO [BomUploadProcessingTaskV2] Consumed 1339 components (1589 before de-duplication), 0 services (0 before de-duplication), and 350 dependency graph entries [bomSerialNumber=08e08436-c806-4460-85b9-301221203475, bomFormat=CycloneDX, bomUploadToken=f451988e-4d54-4f3e-9555-d3853f07ff19, projectName=__redacted__, bomSpecVersion=1.6, projectUuid=4e0d24b4-b3f1-45a2-9489-c604fc0b4aa6, projectVersion=__redacted__, bomVersion=1]
2024-08-02 07:43:45,036 INFO [BomUploadProcessingTaskV2] Processing 1339 components [bomSerialNumber=08e08436-c806-4460-85b9-301221203475, bomFormat=CycloneDX, bomUploadToken=f451988e-4d54-4f3e-9555-d3853f07ff19, projectName=__redacted__, bomSpecVersion=1.6, projectUuid=4e0d24b4-b3f1-45a2-9489-c604fc0b4aa6, projectVersion=__redacted__, bomVersion=1]
2024-08-02 07:43:45,212 INFO [ProjectResource] Project __redacted__ : __redacted__ : __redacted__ updated by odt_****************************MSNV
2024-08-02 07:43:47,301 ERROR [BomUploadProcessingTaskV2] Failed to process BOM [bomUploadToken=f451988e-4d54-4f3e-9555-d3853f07ff19, projectName=__redacted__, projectUuid=4e0d24b4-b3f1-45a2-9489-c604fc0b4aa6, projectVersion=__redacted__]
javax.jdo.JDOUserException: The query returned more than one instance BUT either unique is set to true or only aggregates are to be returned, so should have returned one result maximum
	at org.datanucleus.api.jdo.JDOAdapter.getJDOExceptionForNucleusException(JDOAdapter.java:698)
	at org.datanucleus.api.jdo.JDOQuery.executeInternal(JDOQuery.java:456)
	at org.datanucleus.api.jdo.JDOQuery.executeUnique(JDOQuery.java:360)
	at org.dependencytrack.persistence.LicenseQueryManager.getLicenseByIdOrName(LicenseQueryManager.java:102)
	at org.dependencytrack.persistence.QueryManager.getLicenseByIdOrName(QueryManager.java:624)
	at java.base/java.util.HashMap.computeIfAbsent(Unknown Source)
	at org.dependencytrack.tasks.BomUploadProcessingTaskV2.resolveAndApplyLicense(BomUploadProcessingTaskV2.java:685)
	at org.dependencytrack.tasks.BomUploadProcessingTaskV2.processComponents(BomUploadProcessingTaskV2.java:399)
	at org.dependencytrack.tasks.BomUploadProcessingTaskV2.lambda$processBom$0(BomUploadProcessingTaskV2.java:301)
	at org.dependencytrack.persistence.QueryManager.lambda$runInTransaction$0(QueryManager.java:1433)
	at org.dependencytrack.persistence.QueryManager.runInTransaction(QueryManager.java:1464)
	at org.dependencytrack.persistence.QueryManager.runInTransaction(QueryManager.java:1432)
	at org.dependencytrack.tasks.BomUploadProcessingTaskV2.processBom(BomUploadProcessingTaskV2.java:296)
	at org.dependencytrack.tasks.BomUploadProcessingTaskV2.processEvent(BomUploadProcessingTaskV2.java:187)
	at org.dependencytrack.tasks.BomUploadProcessingTaskV2.inform(BomUploadProcessingTaskV2.java:162)
	at alpine.event.framework.BaseEventService.lambda$publish$0(BaseEventService.java:110)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: org.datanucleus.store.query.QueryNotUniqueException: The query returned more than one instance BUT either unique is set to true or only aggregates are to be returned, so should have returned one result maximum
	at org.datanucleus.store.query.Query.executeQuery(Query.java:2039)
	at org.datanucleus.store.query.Query.executeWithMap(Query.java:1911)
	at org.datanucleus.api.jdo.JDOQuery.executeInternal(JDOQuery.java:437)
	... 17 common frames omitted
```

The uploaded BOM represents a container image and is quite big. However, I checked the licenses and could not find any duplicate license with the same `name` within the same `licenses` node.

--- a/src/main/java/org/dependencytrack/persistence/ComponentQueryManager.java
+++ b/src/main/java/org/dependencytrack/persistence/ComponentQueryManager.java
@@ -863,10 +863,31 @@ final class ComponentQueryManager extends QueryManager implements IQueryManager 
 
         // Group properties by group, name, and value. Because CycloneDX supports duplicate
         // property names, uniqueness can only be determined by also considering the value.
+        final var existingPropertyIdentitiesSeen = new HashSet<ComponentProperty.Identity>();
+        final var existingDuplicateProperties = new HashSet<ComponentProperty>();
         final var existingPropertiesByIdentity = component.getProperties().stream()
+                // The legacy BOM processing in <= 4.11.x allowed duplicates to be persisted.
+                // Collectors#toMap fails upon encounter of duplicate keys.
+                // Prevent existing duplicates from breaking this.
+                // https://github.com/DependencyTrack/dependency-track/issues/4027
+                .filter(property -> {
+                    final var identity = new ComponentProperty.Identity(property);
+                    final boolean isUnique = existingPropertyIdentitiesSeen.add(identity);
+                    if (!isUnique) {
+                        existingDuplicateProperties.add(property);
+                    }
+
+                    return isUnique;
+                })
                 .collect(Collectors.toMap(ComponentProperty.Identity::new, Function.identity()));
+        final var incomingPropertyIdentitiesSeen = new HashSet<ComponentProperty.Identity>();
         final var incomingPropertiesByIdentity = properties.stream()
+                .filter(property -> incomingPropertyIdentitiesSeen.add(new ComponentProperty.Identity(property)))
                 .collect(Collectors.toMap(ComponentProperty.Identity::new, Function.identity()));
+
+        if (!existingDuplicateProperties.isEmpty()) {
+            pm.deletePersistentAll(existingDuplicateProperties);
+        }
 
         final var propertyIdentities = new HashSet<ComponentProperty.Identity>();
         propertyIdentities.addAll(existingPropertiesByIdentity.keySet());

--- a/src/test/java/org/dependencytrack/tasks/BomUploadProcessingTaskTest.java
+++ b/src/test/java/org/dependencytrack/tasks/BomUploadProcessingTaskTest.java
@@ -53,6 +53,7 @@ import org.junit.After;
 import org.junit.Before;
 import org.junit.Test;
 
+import javax.jdo.JDOObjectNotFoundException;
 import java.nio.charset.StandardCharsets;
 import java.time.Duration;
 import java.util.ArrayList;
@@ -65,6 +66,7 @@ import java.util.concurrent.ConcurrentLinkedQueue;
 
 import static org.apache.commons.io.IOUtils.resourceToByteArray;
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
 import static org.assertj.core.api.Assertions.assertThatNoException;
 import static org.assertj.core.api.Assertions.fail;
 import static org.awaitility.Awaitility.await;
@@ -976,6 +978,72 @@ public class BomUploadProcessingTaskTest extends PersistenceCapableTest {
             assertThat(property.getPropertyName()).isEqualTo("foo");
             assertThat(property.getPropertyValue()).isEqualTo("baz");
             assertThat(property.getUuid()).isNotEqualTo(componentProperty.getUuid());
+        });
+    }
+
+    @Test
+    public void informWithExistingDuplicateComponentPropertiesAndBomWithDuplicateComponentProperties() {
+        final var project = new Project();
+        project.setName("acme-app");
+        qm.persist(project);
+
+        final var component = new Component();
+        component.setProject(project);
+        component.setName("acme-lib");
+        component.setClassifier(Classifier.LIBRARY);
+        qm.persist(component);
+
+        final var componentPropertyA = new ComponentProperty();
+        componentPropertyA.setComponent(component);
+        componentPropertyA.setPropertyName("foo");
+        componentPropertyA.setPropertyValue("bar");
+        componentPropertyA.setPropertyType(PropertyType.STRING);
+        qm.persist(componentPropertyA);
+
+        final var componentPropertyB = new ComponentProperty();
+        componentPropertyB.setComponent(component);
+        componentPropertyB.setPropertyName("foo");
+        componentPropertyB.setPropertyValue("bar");
+        componentPropertyB.setPropertyType(PropertyType.STRING);
+        qm.persist(componentPropertyB);
+
+        final var bomUploadEvent = new BomUploadEvent(qm.detach(Project.class, project.getId()), """
+                {
+                  "bomFormat": "CycloneDX",
+                  "specVersion": "1.4",
+                  "serialNumber": "urn:uuid:3e671687-395b-41f5-a30f-a58921a69b79",
+                  "version": 1,
+                  "components": [
+                    {
+                      "type": "library",
+                      "name": "acme-lib",
+                      "properties": [
+                        {
+                          "name": "foo",
+                          "value": "bar"
+                        },
+                        {
+                          "name": "foo",
+                          "value": "bar"
+                        }
+                      ]
+                    }
+                  ]
+                }
+                """.getBytes());
+        new BomUploadProcessingTask().inform(bomUploadEvent);
+        awaitBomProcessedNotification(bomUploadEvent);
+
+        qm.getPersistenceManager().evictAll();
+        assertThatNoException()
+                .isThrownBy(() -> qm.getPersistenceManager().refresh(componentPropertyA));
+        assertThatExceptionOfType(JDOObjectNotFoundException.class)
+                .isThrownBy(() -> qm.getPersistenceManager().refresh(componentPropertyB));
+        assertThat(component.getProperties()).satisfiesExactly(property -> {
+            assertThat(property.getGroupName()).isNull();
+            assertThat(property.getPropertyName()).isEqualTo("foo");
+            assertThat(property.getPropertyValue()).isEqualTo("bar");
+            assertThat(property.getUuid()).isEqualTo(componentPropertyA.getUuid());
         });
     }
 

